book/src/pil/expressions.md

```diff
@@ -0,0 +1,66 @@
+# Expressions
+
+Depending on the context, powdr allows more or less features for expressions.
+
+Inside values for declarations, you can use a very flexible language which includes
+many different operators, function calls, lambda functions, tuple types, statement blocks,
+match statements and others.
+
+In statements and expressions that evaluate to polynomial identities, only a much more restrictive
+language can be used. Expressions in that language are caleld Algebraic Expressions. While you can use
+the full language everywhere, in the context of a polynomial identity, the result after function evaluation
+and constant propagation has to be an algebraic expression.
+
+## Generic Expressions
+
+The expression language allows the following operators, in order of increased precedence:
+
+- lambda functions: ``|params| body``. Examples: ``|i| i`` (the identity), ``|a, b| a + b`` (sum)
+- ``||`` - logical or
+- ``&&`` - logical and
+- ``<``, ``<=``, ``==``, ``!=``, ``>=``, ``>`` - comparisons
+- ``|`` - bitwise or
+- ``^`` - bitwise xor
+- ``&`` - bitwise and
+- ``<<``, ``>>`` - left and right shift
+- ``+``, ``-`` - addition and subtraction (binary operator)
+- ``*``, ``/``, ``%`` - multiplication, division and modulo
+- ``**`` - exponentiation
+- ``-``, ``!`` - numerical and logical negation (unary operators, prefix)
+- ``'`` - "next row" operator (suffix)
+- ``[]``, ``()`` - array index access and function calls
+
+Elementary expressions are
+- number literals (integers)
+- string literals, written in double quotes, e.g. ``"hello"``
+- array literals written in square brackets, e.g. ``[1, 2, 3]``
+- tuples, having at least two elements, e.g. `(1, "abc")`
+- match expressions (see below).
+
```

Comment:
What about the fixed array operators?

Reply:
they're not mentioned here

Reply:
I mean like `[1, 0, 0] + [0, 1]*`

Reply:
Oh these are not part of the expression language. I mean with allowing `+` for array concatenation, they kind of are, but iteration is not.

Reply:
I don't understand why not. If we mention array literals here, aren't these just a way to represent array literals?

Reply:
`*` is not implemented for array literals. What you are referring to is a special syntax to declare fixed columns. I don't see how to especially implement the `*` in the generic expression syntax, since the result is not really an array type (you cannot determine the length without context). It would be better to evaluate those to functions instead of arrays (in the end, fixed columns are also functions and not arrays).

```
x + y => |i| if i < x.len { x(i) } else { y(i - x.len) }
x* => |i| x(i % x.len)
```

Note that both operators need a `.len` property on the input, so they actually should also return such a property, but for `*` it does not exist.

Reply:
ok